src/rulesets/Base/mapreduce.jl

```diff
@@ -44,7 +47,11 @@ function rrule(
 ) where {T<:Union{Real,Complex}}
     y = sum(abs2, x; dims=dims)
     function sum_abs2_pullback(ȳ)
-        return (NO_FIELDS, DoesNotExist(), 2 .* real.(ȳ) .* x)
+        x_thunk = InplaceableThunk(
+            @thunk(2 .* real.(ȳ) .* x),
+            dx -> dx .+= 2 .* real.(ȳ) .* x
```

Comment:
Is this more performant then 
`dx -> dx .+= val` (where `val` is pre-computed as `val = 2 .* real.(ȳ) .* x`) ? (something to do with SIMD?)

Otherwise i gather (from the docs) we could just use a regular `Thunk`. From the docs:
> `add!` should be defined such that: `ithunk.add!(Δ) = Δ .+= ithunk.val` but **it should do this more efficently than simply doing this directly.**

Reply:
The point is to avoid allocating another array the same size as `x`, when there is an existing one we can accumulate on top of.

Reply:
isn't that what would happen anyway? 

i don't know, i'm just going by the docs which implies that `Δ .+= ithunk.val` with the `.+=` is what would happen anyway... or _should_ happen anyway, since i guess this is something an AD would do.

Maybe i don't understand the docs. I think @oxinabox wrote them, so may wish to comment?

Reply:
oh, maybe this is to avoid allocating an intermediate `val = 2 .* real.(ȳ) .* x` array? 

I'd guess a lot of our `Thunks` return arrays... i wonder if writing something like this (where the array computating gets repeated twice) is common enough that we would benefit from having a macro for it (to save on code/typos) e.g. so we just need to write`@inplaceable_thunk(2 .* real.(ȳ) .* x)`?

Reply:
Yes, the broadcast only gets fused if it's all one expression. Otherwise it makes an intermediate. 

A macro like `@inplaceable_thunk` would be pretty neat. This could simply be the default setting of `@thunk` too, I guess, since I can't imagine returning a broadcasting expression which it wouldn't then pay to fuse.

Reply:
made an issue attempting to summarise this discussion :) https://github.com/JuliaDiff/ChainRulesCore.jl/issues/274

Reply:
Nice, thanks. This is not blocking for this PR.
We can come back and edit this code once we resolve that issue.